Spark 3.4 Support

Is it planned to support spark 3.4.x ?

There is some incompatibility on spark 3.5
```

Caused by: java.lang.NoSuchMethodException: org.apache.spark.sql.avro.AvroDeserializer.<init>(org.apache.avro.Schema, org.apache.spark.sql.types.DataType, java.lang.String)
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.getConstructor(Class.java:1825)
	at org.apache.spark.sql.avro.AbrisAvroDeserializer$$anonfun$1.applyOrElse(AbrisAvroDeserializer.scala:38)
	at org.apache.spark.sql.avro.AbrisAvroDeserializer$$anonfun$1.applyOrElse(AbrisAvroDeserializer.scala:37)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at scala.util.Failure.recover(Try.scala:234)
	at org.apache.spark.sql.avro.AbrisAvroDeserializer.<init>(AbrisAvroDeserializer.scala:37)
	at za.co.absa.abris.avro.sql.AvroDataToCatalyst.deserializer$lzycompute(AvroDataToCatalyst.scala:71)
	at za.co.absa.abris.avro.sql.AvroDataToCatalyst.deserializer(AvroDataToCatalyst.scala:71)
	at za.co.absa.abris.avro.sql.AvroDataToCatalyst.nullSafeEval(AvroDataToCatalyst.scala:87)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenEvaluatorFactory$WholeStageCodegenPartitionEvaluator$$anon$1.hasNext(WholeStageCodegenEvaluatorFactory.scala:43)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at scala.collection.convert.Wrappers$IteratorWrapper.hasNext(Wrappers.scala:32)
	at com.google.common.collect.Iterators$PeekingImpl.hasNext(Iterators.java:1139)
	at com.databricks.photon.NativeRowBatchIterator.hasNext(NativeRowBatchIterator.java:44)
	at 0xa8e0d62 <photon>.HasNext(external/workspace_spark_3_5/photon/jni-wrappers/jni-row-batch-iterator.cc:50)
	at 0x5e136fb <photon>.OpenImpl(external/workspace_spark_3_5/photon/exec-nodes/file-writer-node.cc:166)
	at com.databricks.photon.JniApiImpl.open(Native Method)
	at com.databricks.photon.JniApi.open(JniApi.scala)
	at com.databricks.photon.JniExecNode.open(JniExecNode.java:71)
	at com.databricks.photon.PhotonWriteResultHandler.$anonfun$getResult$3(PhotonWriteStageExec.scala:121)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at com.databricks.photon.PhotonResultHandler.timeit(PhotonResultHandler.scala:30)
	at com.databricks.photon.PhotonResultHandler.timeit$(PhotonResultHandler.scala:28)
	at com.databricks.photon.PhotonWriteResultHandler.timeit(PhotonWriteStageExec.scala:67)
	at com.databricks.photon.PhotonWriteResultHandler.$anonfun$getResult$2(PhotonWriteStageExec.scala:121)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1542)
	at com.databricks.photon.PhotonWriteResultHandler.getResult(PhotonWriteStageExec.scala:118)
	... 38 more
```

--- a/src/main/scala/org/apache/spark/sql/avro/AbrisAvroDeserializer.scala
+++ b/src/main/scala/org/apache/spark/sql/avro/AbrisAvroDeserializer.scala
@@ -20,6 +20,7 @@ import org.apache.avro.Schema
 import org.apache.spark.sql.types.DataType
 import za.co.absa.commons.annotation.DeveloperApi
 
+import scala.collection.mutable
 import scala.util.Try
 
 /**
@@ -31,18 +32,37 @@ class AbrisAvroDeserializer(rootAvroType: Schema, rootCatalystType: DataType) {
 
   private val deserializer = {
     val clazz = classOf[AvroDeserializer]
-    Try {
-      clazz.getConstructor(classOf[Schema], classOf[DataType])
-        .newInstance(rootAvroType, rootCatalystType) // Spark 2.4 -
-    }.recover { case _: NoSuchMethodException =>
-      clazz.getConstructor(classOf[Schema], classOf[DataType], classOf[String])
-        .newInstance(rootAvroType, rootCatalystType, "LEGACY") // Spark 3.0 - Spark 3.5.0 (including)
-    }.recover { case _: NoSuchMethodException =>
-      clazz.getConstructor(classOf[Schema], classOf[DataType], classOf[String], classOf[Boolean])
-        .newInstance(rootAvroType, rootCatalystType, "LEGACY", false: java.lang.Boolean) // Spark 3.5.x +
+    val schemaClz = classOf[Schema]
+    val dataTypeClz = classOf[DataType]
+    val stringClz = classOf[String]
+    val booleanClz = classOf[Boolean]
+
+    clazz.getConstructors.collectFirst {
+      case currCtor if currCtor.getParameterTypes sameElements
+        Array(schemaClz, dataTypeClz) =>
+        // Spark 2.4
+        currCtor.newInstance(rootAvroType, rootCatalystType)
+      case currCtor if currCtor.getParameterTypes sameElements
+        Array(schemaClz, dataTypeClz, stringClz) =>
+        // Spark 3.0 - Spark 3.5.0 (including)
+        currCtor.newInstance(rootAvroType, rootCatalystType, "LEGACY")
+      case currCtor if currCtor.getParameterTypes sameElements
+        Array(schemaClz, dataTypeClz, stringClz, booleanClz) =>
+        // Spark 3.5.1 - 3.5.2
+        currCtor.newInstance(rootAvroType, rootCatalystType, "LEGACY", false: java.lang.Boolean)
+      case currCtor if currCtor.getParameterTypes.toSeq sameElements
+        Array(schemaClz, dataTypeClz, stringClz, booleanClz, stringClz) =>
+        // Spark 4.0.0-SNAPSHOT+
+        currCtor.newInstance(rootAvroType, rootCatalystType, "LEGACY", false: java.lang.Boolean, "")
+    } match {
+      case Some(value: AvroDeserializer) =>
+        value
+      case _ =>
+        throw new NoSuchMethodException(
+          s"""Supported constructors for AvroDeserializer are:
+             |${clazz.getConstructors.toSeq.mkString(System.lineSeparator())}""".stripMargin)
     }
-      .get
-      .asInstanceOf[AvroDeserializer]
+
   }
 
   private val ru = scala.reflect.runtime.universe

